Editor: Form Builder Dialog

rodrigoi opened the pull request:
As part of implementing support for contact formson the editor, a new dialog window has been implemented to host a building interface.

A little reminder of cloudup's editor implementation:

![cloudup](https://cloud.githubusercontent.com/assets/233601/11937254/e4357ea0-a7f1-11e5-89d4-9be977bb06b7.png)

The proposed solution uses a `FoldableCard` component for each of the form fields. 
The card displays the field name and a human readable description of the field. Also, a remove button is added to the card header to simplify removal of fields. Otherwise, the user would have to expand the card and click on a hidden _remove_ button.

![form fields](https://cloud.githubusercontent.com/assets/233601/11937261/eb43753a-a7f1-11e5-9d85-b272eac46b8c.png)

This is the proposed view for the expanded card for a text field:
![text field](https://cloud.githubusercontent.com/assets/233601/11937274/fda046ea-a7f1-11e5-9060-0404eb342ffa.png)

And this is the proposed view for a radio button list:
![radio buttons](https://cloud.githubusercontent.com/assets/233601/11937265/f2e59124-a7f1-11e5-84e9-2021dabef778.png)

Code is not complete. When design is agreed this PR will be updated for code review.

Changing the order of the form fields and form preview will be reviewed in future pull requests.

This PR is part of #1404 and depends on #2069.

drw158 commented:
Overall I think the design and the components you chose are headed in the right direction.

- A new contact form should include a simple form to start out with (Name, Email, Subject, Message).
- Overall, this feels uncomfortable this wide. It should be narrower.
- We can keep the foldable card with the button, but I don't think the sub text is necessary.
- The names of the elements in ContactFormDialogFieldHeader could be more descriptive. For example instead of "Name" use "Name textfield" or "Name textbox". Maybe use the form label that they chose as well? I can help with the new names.
- We should just use a borderless red trash icon here instead of the red primary Remove button.

In the expanded card:
- I changed "Field Name"  to"Field Label". https://github.com/Automattic/wp-calypso/commit/77444cb2b1c2bb09a64acbcd18142559fc69d6ba
- I tweaked the descriptive text a little bit. https://github.com/Automattic/wp-calypso/commit/77444cb2b1c2bb09a64acbcd18142559fc69d6ba
- The required option is using too much text I think. It should just read "Required" with no label.

- Add New Field should be a regular button, not primary.

The big issue:
 
These field should be draggable so that the user can move them around. Unfortunately, we don't have a good UI for that in Calypso that I know of. I think it should work similarly to this:

![reorder](https://cloud.githubusercontent.com/assets/618551/12122615/ca933332-b3a1-11e5-9bc1-82f20700a583.gif)

I believe that we can get a good experience like this on mobile web too, although a good drag and drop experience is hard to find on mobile web :).

One last thing to think about: should the name field have a first and last name?

drw158 commented:
Ah, sorry, it seems that the ContactFormDialogFieldHeader already uses the label as the text.

roundhill commented:
Design looks good to me. @rodrigoi are you merging this up now that #2069 has been merged? Happy to test it out once it's ready.

mtias commented:
@drw158 we have <a href="https://github.com/Automattic/wp-calypso/tree/master/client/components/forms/sortable-list">SortableList</a> that could be improved upon if you need to.

rodrigoi commented:
This is ready for review and testing :smile: 

What is **not** included on this PR:
- Contact form preview. This PR covers changes on the dialog window.
- Drag and Drop. A `SortableList` implementation is in the works.

### Summary of Changes:

#### Configuration
This PR enables this feature on _WPCalypso_ and _Stage_ for a wider testing when merged.

#### Fixes
The field types were changed to match the legacy contact form. Dropdown type is `select` and web address is now `url`.

The shortcode serializer/deserializer has been updated to work correctly with fields that have a list of options. This includes unit tests.

#### State Management
Two more actions/reducers were added with their corresponding tests to support updating an individual field and changing the form settings.

#### UI
The provisional `dialog.jsx` has been replaced by five components that handle navigation within the dialog, settings, fields, individual fields and the header within the `FoldableCard`. One utility module exports a function to handle the localization of field types labels.

The toolbar button is text only (matching the legacy editor style). I didn't find a suitable gridicon for this.

![toolbar button](https://cloud.githubusercontent.com/assets/233601/13006928/213c7306-d16b-11e5-9bb1-a267594f3a22.png)

And the same default form is presented to the user.

![default form](https://cloud.githubusercontent.com/assets/233601/13006914/095beb9a-d16b-11e5-85b8-6b2499537ff4.png)

Fields are edited using a `FoldableCard`

![edit field](https://cloud.githubusercontent.com/assets/233601/13006912/095b7728-d16b-11e5-9dca-a6a61863c0db.png)

And options for radio button lists and dropdowns are handled using a `TokenField`.

![edit field with options](https://cloud.githubusercontent.com/assets/233601/13006915/095c416c-d16b-11e5-9395-081231935cdc.png)

The settings tab uses the same copy as the legacy editor. These texts may need change.

![form settings](https://cloud.githubusercontent.com/assets/233601/13006913/095b7142-d16b-11e5-87bc-0accbb1ac894.png)

@drw158 it'll be great if you can sign off the design and the copy texts.

@mtias, @aduth. @artpi and @gwwar, you've all reviewed some previous stage of this. I'll be awesome if you can take a look at the code.

drw158 commented:
@rodrigoi Before I dig deep, I think the Add Contact Form button should be an icon (no text), located right after the Read More button. @mtias suggested the `@` icon in Gridicons and I agree.

rodrigoi commented:
@drw158 updated!

![screen shot 2016-02-12 at 14 11 53](https://cloud.githubusercontent.com/assets/233601/13014532/14c8c444-d193-11e5-8706-5013b0e87b95.png)

drw158 commented:
- To be consistent with the media uploader, the first primary action should be `Insert`, not Save. If you edit the form later, the primary action should change to  `Update`.
- Description text for the `Options` field: "Insert an option and press enter."
- We should pre-fill the option field with 2 options for better clarity, [like so](https://cloudup.com/cH5P_L3yJhF).
- Instead of `Radio` we should use `Radio Button` instead. (I think that is a more common term)
- I feel like the pencil should have an active state. After you click it, it should stay blue while you are editing.

Form settings:

![screen shot 2016-02-12 at 2 55 00 pm](https://cloud.githubusercontent.com/assets/618551/13020507/690c610c-d19b-11e5-8239-4adcf8835e82.png)

Note that the fields already have the user's info (author's email and the title of the post). If it's too complicated to pre-fill the fields with that info, then I have some alternate designs with different description text.

Bugs:

- If I add a new field, change it to Radio, insert some options, and then change the type to Web Address, the descriptive text reads `Optional Web Address with 3 options`.

drw158 commented:
@rodrigoi I think the `@` icon works well, much better. It should be placed right after the read more button icon if possible, though.

gwwar commented on client/components/tinymce/plugins/contact-form/dialog/field-list.jsx:
```diff
@@ -0,0 +1,47 @@
+
+		return <EmptyContent
+			title={ null }
+			line={ this.translate( 'An empty form is no fun!. Go ahead and add some fields!' ) }
```
Extra period here

gwwar commented on client/components/tinymce/plugins/contact-form/dialog/field-list.jsx:
```diff
@@ -0,0 +1,47 @@
+							<Field
+								key={ index }
+								{ ...{ label, type, options, required } }
+								onRemove={ () => this.props.onRemoveField( index ) }
```
We're trigging the eslint rule: https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/jsx-no-bind.md
For performance reasons we want to avoid bind/fat arrow. See: p4TIVU-2nh-p2

In this case, we should be able to avoid this. Pass through index as a prop to Field, and in Field.jsx add a method like the following:
```javascript
handleRemoveClick( event ) {
     this.props.onRemoveField( event, this.props.index );
}
```
and
```javascript
const remove = <Gridicon icon="trash" onClick={ this.handleRemoveClick } className="editor-contact-form-modal-field__remove" />;
```

Same deal with onUpdate below.

gwwar commented on client/components/tinymce/plugins/contact-form/dialog/field.jsx:
```diff
@@ -0,0 +1,100 @@
+					<FormLabel>{ this.translate( 'Options' ) }</FormLabel>
+					<TokenField
+						value={ options || [] }
+						onChange={ tokens => this.props.onUpdate( { options: tokens.join() } ) }/>
```
Move this to a static function, as we recreate this function on each render.

gwwar commented on client/components/tinymce/plugins/contact-form/dialog/field.jsx:
```diff
@@ -0,0 +1,100 @@
+					<FormLabel>
+						<FormCheckbox
+							checked={ this.props.required }
+							onChange={ () => this.props.onUpdate( { required: ! this.props.required } ) } />
```
This can be moved to a named function on the class.

gwwar commented on client/components/tinymce/plugins/contact-form/dialog/field.jsx:
```diff
@@ -0,0 +1,100 @@
+							<DropdownItem
+								key={ 'field-type-' + fieldType }
+								selected={ this.props.type === fieldType }
+								onClick={ () => this.props.onUpdate( { type: fieldType } ) }>
```
If you can, try to see if you can avoid the fat arrow here.

gwwar commented on client/state/action-types.js:
```diff
@@ -17,6 +17,8 @@ export const EDITOR_CONTACT_FORM_ADD_DEFAULT_FIELD = 'EDITOR_CONTACT_FORM_ADD_DE
 export const EDITOR_CONTACT_FORM_LOAD_FORM = 'EDITOR_CONTACT_FORM_LOAD_FORM';
 export const EDITOR_CONTACT_FORM_REMOVE_FIELD = 'EDITOR_CONTACT_FORM_REMOVE_FIELD';
+export const EDITOR_CONTACT_FORM_UPDATE_FIELD = 'EDITOR_CONTACT_FORM_UPDATE_FIELD';
+export const EDITOR_CONTACT_FORM_UPDATE_SETTINGS = 'EDITOR_CONTACT_FORM_UPDATE_SETTINGS';
```
As per https://wpcalypso.wordpress.com/devdocs/docs/our-approach-to-data.md#actions we have a preference for the verb to be suffixed on action names. Related actions should probably be renamed at the same time, so I'm OK with it not making it into this PR.

So `EDITOR_CONTACT_FORM_FIELD_UPDATE` vs `EDITOR_CONTACT_FORM_UPDATE_FIELD`

gwwar commented on client/state/ui/editor/contact-form/test/reducer.js:
```diff
@@ -171,4 +173,122 @@ describe( "editor's contact form state reducer", () => {
+			const state = reducer( contactForm, {
+				type: EDITOR_CONTACT_FORM_UPDATE_FIELD,
+				index: 2,
+				field: { label: 'Web Address', type: 'url', required: true }
```
Optional: use `deepFreeze` to detect mutations on contactForm

aduth replied:
>Optional: use `deepFreeze` to detect mutations on contactForm

Required in my opinion :smile: At minimum, very strongly recommended. It's caught me a few times.

gwwar commented:
:+1: Looking pretty slick! Codewise this is good to :ship: after some of the eslint warnings are fixed.

aduth commented on client/components/tinymce/plugins/contact-form/dialog/locales.js:
```diff
@@ -0,0 +1,18 @@
+}
+
+export default function ( type ) {
+	return type in labels
```
For safety, you should probably use `labels.hasOwnProperty( type )` here. Try `locales( 'constructor' );` as an example to see what I mean.

aduth commented on client/components/tinymce/plugins/contact-form/dialog/locales.js:
```diff
@@ -0,0 +1,18 @@
+
+export default function ( type ) {
+	return type in labels
+		? labels[type].call()
```
No ESLint warning (because of exceptions), but there should be spaces between the square brackets here ([reference](https://github.com/Automattic/wp-calypso/blob/master/docs/coding-guidelines/javascript.md#spacing)).

aduth commented on client/components/tinymce/plugins/contact-form/dialog/locales.js:
```diff
@@ -0,0 +1,18 @@
+	url: () => i18n.translate( 'Web Address' )
+}
+
+export default function ( type ) {
```
ESLint warning: Drop the space before the opening paren.

aduth commented on client/components/tinymce/plugins/contact-form/dialog/field-header.jsx:
```diff
@@ -0,0 +1,53 @@
+				} );
+		}
+
+		return this.translate( '%(required)s %(type)s', {
```
Consider whether these strings are comprehensible without context. It may make sense to add a `comment` for translators.

aduth commented on client/components/tinymce/plugins/contact-form/dialog/field.jsx:
```diff
@@ -0,0 +1,100 @@
+		return (
+			<FoldableCard
+				header={ <FieldHeader { ...omit( this.props, [ 'onRemove', 'onUpdate' ] ) } /> }
+				icon={ 'pencil' }
```
A consistency point, since there's mixed usage here, but this can be `icon="pencil"`

aduth commented on client/components/tinymce/plugins/contact-form/dialog/field.jsx:
```diff
@@ -0,0 +1,100 @@
+			);
+		}
+
+		return null;
```
Two points on this function. First, it generally helps readability if you return early whenever possible, as it avoids the majority of the function being indented an extra level. Second, only the `render` method requires that a non-`undefined` value be returned, so `return null;` could be omitted altogether here (leveraging that functions without `return` implicitly `return undefined;`).

I think this function should be rewritten as:

```js
renderOptions() {
	if ( this.props.type !== 'radio' && this.props.type !== 'select' ) {
		return;
	}

	// ...

	return (
		<FormFieldSet>
			{ /* ... */ }
		</FormFieldSet>;
	);
}
```

aduth commented on client/components/tinymce/plugins/contact-form/dialog/field.jsx:
```diff
@@ -0,0 +1,100 @@
+	renderOptions() {
+		if ( this.props.type === 'radio' || this.props.type === 'select' ) {
+			let { options } = this.props;
+			if ( options && typeof options === 'string' ) {
```
It's not clear to me when `options` would be truthy and not a string?

aduth commented on client/components/tinymce/plugins/contact-form/dialog/index.jsx:
```diff
@@ -1,13 +1,98 @@
+	displayName: 'ContactFormDialog',
+
+	propTypes: {
+		activeTab: PropTypes.oneOf( ['fields', 'settings'] ).isRequired,
```
Arrays should have spaces between square brackets ([reference](https://github.com/Automattic/wp-calypso/blob/master/docs/coding-guidelines/javascript.md#spacing))

aduth commented on client/components/tinymce/plugins/contact-form/dialog/index.jsx:
```diff
@@ -1,13 +1,98 @@
+
+		if ( this.props.activeTab === 'fields' ) {
+			return [
+				<div className="editor-contact-form-modal__secondary-actions">
```
Does this not need a `key`? I'm not seeing a warning in the console, so perhaps not.

aduth commented on client/components/tinymce/plugins/contact-form/dialog/index.jsx:
```diff
@@ -1,13 +1,98 @@
+			onUpdateSettings
+		} = this.props;
+
+		const content = this.props.activeTab === 'fields'
```
You've deconstructed this prop above, so this could be shortened to `const content = activeTab === 'fields'`

aduth replied:
There's also a few other props in this function called via `this.props.` . Seems like we should strive for consistency by either deconstructing all or calling all on `this.props`

aduth commented on client/components/tinymce/plugins/contact-form/dialog/navigation.jsx:
```diff
@@ -0,0 +1,41 @@
+
+	propTypes: {
+		fieldCount: PropTypes.number.isRequired,
+		activeTab: PropTypes.oneOf( ['fields', 'settings'] ).isRequired,
```
Same as https://github.com/Automattic/wp-calypso/pull/1885/files#r52916675

aduth commented on client/components/tinymce/plugins/contact-form/dialog/settings.jsx:
```diff
@@ -0,0 +1,53 @@
+				<SectionHeader label={ this.translate( 'Contact Form Notification Settings' ) } />
+				<Card>
+					<p>
+						{ this.translate( 'If you don’t make any changes here, feedback will be sent to the author of the page/post and the subject will be the name of this page/post.' ) }
```
Elsewhere in the editor, we've tried to create unique strings depending on the post type, rather than imposing the burden of differentiating "post/page" on the user.

May be worth exploring whether it's possible to provide the post type as an argument to the string translation. Having a generalized approach here will be good for when we add support for custom post types.

Could you see if this works?

```js
this.translate( '...feedback will be sent to the author of the %(postType)s', { args: { postType: this.translate( 'post' ) } } );
```

aduth commented on client/components/tinymce/plugins/contact-form/style.scss:
```diff
@@ -0,0 +1,73 @@
+	height: 40px;
+}
+
+button.foldable-card__action.foldable-card__expand .gridicon.gridicons-pencil {
```
Couldn't this style apply outside the context of the editor usage? Seems like it should be scoped to the contact form.

aduth commented on client/state/ui/editor/contact-form/reducer.js:
```diff
@@ -26,14 +31,21 @@ export default function( state = initalState, action ) {
 			break;
+		case EDITOR_CONTACT_FORM_UPDATE_FIELD:
+			state = cloneDeep( state );
+			state.fields[index] = Object.assign( {}, state.fields[index], field );
```
Spaces between square brackets ([reference](https://github.com/Automattic/wp-calypso/blob/master/docs/coding-guidelines/javascript.md#spacing)).

aduth commented on config/stage.json:
```diff
@@ -18,6 +18,7 @@
 		"post-editor/iframe-preview": true,
 		"post-editor/live-image-updates": true,
 		"post-editor/pages": true,
+		"post-editor/contact-form": true,
```
What's the reasoning for enabling this in staging? My impression of staging is that it's intended for features that are essentially shippable, and very close to/mirroring production readiness.

aduth commented:
First, let me say that this is very cool. I can even envision future iterations of the form abstracted to support general user/variably defined form rendering. In fact, in a current side project I'm looking to support something like this, and have poked around at a few of the existing solutions (e.g. [`react-jsonschema-form`](https://github.com/mozilla-services/react-jsonschema-form)). Anywho, I digress and this is not of any consequence to the pull request here.

Second, I encourage you to keep your pull requests smaller and focused. It's neat to see everything working well as implemented here, but is a challenge to review with the precision that a smaller pull request would enable.

alisterscott commented:
Firstly, this looks great :clap: well done. My feedback is below.

**Design Feedback**

1. The delete and edit icons don't have any hover text like the editor toolbar - this would be good for users to understand what these buttons do as they don't have labels
2. When editing a field clicking the same edit button closes the field's details - this wasn't immediately obvious to me - is there a better icon for closing the field?
3. When adding a new field it would be handy to default to the open state for that field so the user can immediately edit the field without an additional click/tap
4. With options it isn't immediately obvious that the user has to use the 'return' key between entering options like tags -> it would be good to have some explanatory text or use a different widget (a la Google Docs Forms)  ![dropdowns google forms](https://cloud.githubusercontent.com/assets/128826/13068994/4df62d4c-d4cc-11e5-9ec5-31a567e50278.gif)
5. I think "Radio Buttons" would be a more descriptive label than just "Radio"
6. The order of the types is different than wp-admin which is purely alphabetical. Is this intentional? ![wp-admin types](https://cloud.githubusercontent.com/assets/128826/13068985/42be1b4c-d4cc-11e5-87a6-65b3f2f5db04.png) ![calypso types](https://cloud.githubusercontent.com/assets/128826/13068992/476a9e7c-d4cc-11e5-9595-2bc504734729.png)
7. For a new form (not editing existing) would the action "Insert" be clearer than "Save"
8. Would Short Text/Long Text be better descriptions than Text/Text Area? Google Forms uses Short answer/Paragraph

**Bugs**
1. Changing a field from one with options (eg. dropdown) to one without options (eg. name) still displays the options
![name with options](https://cloud.githubusercontent.com/assets/128826/13068946/e77657fe-d4cb-11e5-84ba-9c3cf83f59a7.png)
2. Narrow screens lose the tabs at the top: 
![contact form builder narrow screens](https://cloud.githubusercontent.com/assets/128826/13069045/b1d65d64-d4cc-11e5-8454-7a42cda10afc.png)

ryanboren commented:
https://test.wordpress.com/2016/02/16/editor-forms-using-the-form-builder-dialog-to-create-an-about-page-contact-form-iphone-6/

lancewillett commented:
Found a bug during a test run with IE 11 (Windows 10) — Chrome (Windows 10) didn't have the issue.

![ie11-contact-form-double](https://cloud.githubusercontent.com/assets/66797/13095513/66fb3d7a-d4cf-11e5-86a8-bdbf7109211d.gif)

When you edit a contact form, add a field, and save it again — a double contact form is added to the post editor. Expected: just save and keep the one contact form.

lancewillett commented:
Two issues to report during an i18n test: using Spanish for both UI and site settings; test site https://lancetest35.wordpress.com/ and user `lancetest35`.

1. Not all strings are i18n-ized in the contact form

<img width="1264" alt="screen shot 2016-02-16 at 15 36 29" src="https://cloud.githubusercontent.com/assets/66797/13095550/b490a82c-d4cf-11e5-8e9d-91cf54021122.png">
<img width="1264" alt="screen shot 2016-02-16 at 15 36 47" src="https://cloud.githubusercontent.com/assets/66797/13095551/b4918d96-d4cf-11e5-8c0e-35df28e0085a.png">
<img width="1264" alt="screen shot 2016-02-16 at 15 36 55" src="https://cloud.githubusercontent.com/assets/66797/13095553/b4920b18-d4cf-11e5-814a-d1bc3c5a77e6.png">
<img width="1264" alt="screen shot 2016-02-16 at 15 38 22" src="https://cloud.githubusercontent.com/assets/66797/13095552/b491e3a4-d4cf-11e5-82bf-2c85229ff58b.png">

2. When you submit the contact form, the resulting email has i18n string issues, like date output: `Hora: febrero 17, 2016 am Miércolesam20 12:08 am`

<img width="798" alt="screen shot 2016-02-16 at 15 38 50" src="https://cloud.githubusercontent.com/assets/66797/13095588/0502931a-d4d0-11e5-836c-617652ca9d4a.png">

hoverduck commented:
Not a big problem, since this is also an issue in wp-admin, but as long as you're working on it, it would be nice to have some input validation in a couple places:

1) On Settings->E-mail Address field to ensure a valid e-mail address is used.  And it might also look nice if you used the TokenField component like for the Dropdown/Radio options input.
2) It's possible to define a required Dropdown or Radio field without any options, which makes it impossible to submit the form.

roundhill commented:
Testing with the latest changes worked well for me. I think before this gets any larger and further delayed we should merge it now to get it into testing and do separate PRs for any issues that come up during testing.

mtias commented:
Please make sure to rebase so we have a bit of an easier time reading the commits.

alisterscott commented:
I'd love to see some movement on this as I believe it's a much needed enhancement to the Calypso editor, even if we merged this as is I am happy to raise the issues listed as comments as separate Github tickets for this. If you'd like me to do this, please let me know!

rodrigoi commented:
Hi!

I've updated the code with a lot of changes and it's ready for another review. Here is a summary of what's new or has changed:

@drw158 
- Primary actions are now `Insert` and `Update`.
- Added descriptive text to the options token field.
- Options prefilled with two items.
- _Radio_ changed to _Radio Button_
- The pencil icon remains active when the foldable card is open.
- Now the description label behaves as expected when changing between types with and without options.
- Settings page now uses the current user email and the post title (if available) as placeholders.

@gwwar
- Fixed lint errors.
- Added `deepFreeze` to tests.
- Fixed action types so they are suffixed by the verb.

@aduth
- Fixed lint errors.
- refactored `locales.js`.
- Added comments and context for some translations.
- Fixed consistency issues on `field.jsx`.
- Refactoring of some functions to improve readability.
- The settings page is now aware of the post type and provides correct messages.

@alisterscott 
- Added hover state to delete and edit icons.
- Added descriptive text to the options token field.
- Options prefilled with two items.
- _Radio_ changed to _Radio Button_
- Fixed inconsistencies on the field type list.
- Fixed `z-index` issues on narrow screens.

Please, feel free to create separate issues for things like open state and alternatives to the `TokeField`.

@lancewillett 
Double Form Bug.

I've done some testing and I duplicated this behaviour in Chrome for Mac. Apparently it's an editor general bug. I duplicated the same behaviour using a media gallery.
For example, if the first element on the editor is a contact form preview or a media gallery, when updating the element it gets's duplicated. If you add anything before it (a line break for example), updating the element works just fine.

Internationalization
- Some strings are new and will require translations.
- All issues with the backend are out of the scope of this PR. This plugin uses the same shortcode as the legacy form and the same backend.

@hoverduck 
- Added validation for the emails entered on the settings page. Will evaluate using `TokeField`, but if added it will be on another PR.
- Added validations that prevent adding an invalid form to the document.

gwwar commented on client/components/tinymce/plugins/contact-form/dialog/validations.js:
```diff
@@ -0,0 +1,23 @@
+}
+
+export function validateSettingsToEmail( to ) {
+	const emails = !! to ? to.split( ',' ) : [];
```
Should we trim whitespace here?

gwwar commented on client/state/ui/editor/contact-form/reducer.js:
```diff
@@ -2,37 +2,61 @@
+	const { index, field, settings } = action;
 
-export default function( state = initalState, action ) {
 	switch ( action.type ) {
```
Please, also add in cases for `SERIALIZE` and `DESERIALIZE`. Without them this subtree will persist by default. We can skip persistence by returning an empty object for each:

```javascript
case SERIALIZE:
    return {};
case DESERIALIZE:
   return {};
```

(Also see: https://github.com/Automattic/wp-calypso/blob/master/docs/our-approach-to-data.md#data-persistence--2754-)

aduth replied:
>Please, also add in cases for `SERIALIZE` and `DESERIALIZE`. Without them this subtree will persist by default.

Normally true, except... #3337 :grinning:

gwwar replied:
Oh right! :smile:

gwwar commented on client/state/ui/editor/contact-form/test/reducer.js:
```diff
@@ -2,21 +2,26 @@
 } from 'state/action-types';
 
-describe( "editor's contact form state reducer", () => {
+i18n.initialize();
```
@blowery do we call this in the single test runner?

blowery replied:
we don't currently, but we can add it to the global before suite. It just moved, so this would have to be rebased to pick it up.

blowery replied:
now lives here: https://github.com/Automattic/wp-calypso/blob/master/test/boot-test.js

blowery replied:
another approach is to mock out the i18n code. see https://github.com/Automattic/wp-calypso/pull/4088#discussion_r56437481

We might consider replacing the i18n infrastructure entirely under test...

rodrigoi replied:
@gwwar @blowery I've mocked the `i18n` mixin on https://github.com/Automattic/wp-calypso/pull/1885/commits/6a6405429d8b6c8ba8024a3be6e16797f444c7ed

gwwar commented:
:+1: Due to the size of this PR I may have missed something, but overall this looks and works well. You should be good to :ship: after adding `SERIALIZE` and `DESERIALIZE` handlers in `/ui/editor/contact-form/reducer.js`

I'm sure there are other things we can iterate on, but we should try to land this PR, and follow up with  smaller PR improvements.

alisterscott commented:
Thanks for the update, this is looking great.

Are we planning on updating the placeholder text to say something like "_Embedded Contact Form_" instead of the current "_This is a placeholder for the form preview_"?

Some bugs/enhancements I have raised separately so we can look these after this PR is shipped.

* Bug: double entry is inserted on edit in Chrome - this also impacts galleries so raised generic bug #4115 
* Bug: edit button missing hover text (delete is ok) #4116 
* Enhancement: default to open state for new fields #4117 
* Enhancement: warn if unsaved contact form changes when clicking/touching outside dialog #4118 (this is possibly a bug as wp-admin does this)
* Enhancement: allow leading/trailing space around email address validations #4119 
* Bug: prevent second/subsequent contact forms being added to post/page (as per wp-admin) #4120

aduth commented on client/components/tinymce/plugins/contact-form/dialog/settings.jsx:
```diff
@@ -0,0 +1,78 @@
+							'If you don’t make any changes here, feedback will be sent to the author of the %(postType)s and the subject will be the name of this %(postType)s.',
+							{
+								args: {
+									postType: this.props.postType === 'post' ? this.translate( 'post' ) : this.translate( 'page' )
```
Since my last review, we've added some helpers to make post type labels accessible from within the editor, to our goal of enabling custom post type support in the editor (see #3695).

[`<EditorPostType />` is an example of a component displaying labels for custom post types](https://github.com/Automattic/wp-calypso/blob/master/client/post-editor/editor-post-type/index.jsx). You could probably forgo the hardcoding which remains there for post and page (simply as a convenience for bypassing the loading indicator that would otherwise be visible while the editor loads).

All of this assumes we want contact forms to be enabled for custom post types, so correct me if I'm wrong.

rodrigoi replied:
I'd rather add this on another PR if that's fine.

rodrigoi commented:
@alisterscott awesome! I'll look at those issues when this get's merged. About the placeholder text, there is another PR with those changes: https://github.com/Automattic/wp-calypso/pull/3287

ockham commented:
@rodrigoi This has been in wpcalypso for a while, but never made it to staging or production. Time to turn it on in staging at least?

mtias commented:
@ockham see https://github.com/Automattic/wp-calypso/pull/4545 :)
